[Authentication] Add logic to login via magic link

notandyvee opened the pull request:
### Fix

This PR adds the logic necessary to login via magic link. Instead of entering a password for login, you request an email. You go to the your email and click the login link in order to auto-magically login without password.

### Test

TBD

### Review

Nothing special was added. I avoided a major overhaul. For example, instead of json parsing with a library I used `JSONObject` and used `OkHttp`. See `SimpleHttp`. If we do more with network requests and json parsing it would be worth using retrofit and a better json parsing library, like Moshi. 

### Release
<!--
***(Required)*** Add a concise statement to `RELEASE-NOTES.txt` if the changes should be included in release notes. Include details about updating the notes in this section. For example:
`RELEASE-NOTES.txt` was updated in d3adb3ef with:
> Added markdown support
-->

This change requires release notes.

notandyvee committed:
[Authentication] Add logic to login via magic link

dangermattic commented:
<!--
  0 Errors
  4 Warnings: <code>strings.xml</code> files...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          4 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>strings.xml</code> files should only be updated on release branches, when the translations are downloaded by our automation.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>OkHttpMagicLinkRepository</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>SimpleHttp</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

danilo04 commented on Simplenote/src/main/java/com/automattic/simplenote/DeepLinkActivity.kt:
```diff
@@ -0,0 +1,58 @@
+                startActivity(intent)
+            }
+            LOGIN_SCHEME -> {
+                if (queryParamContainsData(uri.query, "auth_key") && queryParamContainsData(uri.query, "auth_code")) {
```
(⛏️ ) Move `"auth_key"` and `"auth_code"` as constants.

notandyvee committed:
[Bug] Selecting continuing with password works properly.

danilo04 commented on Simplenote/src/main/java/com/automattic/simplenote/authentication/MagicLinkableFragment.kt:
```diff
@@ -0,0 +1,105 @@
+/**
+ * Base class used to share logic between sign up and login, specifically related to magic links.
+ */
+abstract class MagicLinkableFragment : Fragment() {
```
Are you thinking on implementing more fragments related to magic links?

notandyvee replied:
Sorta. This class was based off of `SignupFragment`, which is similar. Both Sign Up and Sign In with magic link are similar. So it made sense to use that logic and share it between them. Additionally, we can likely remove the okhttp use from `SignupFragment` and move it over the business logic to a viewmodel. Hence why I created the `SimpleHttp` helper object. But I didn't want to inflate the work here further.

notandyvee committed:
Add another intent filter to match deeplink for app.simplenote.com/login

notandyvee committed:
Removed transition animation to NotesActivity when validating magic link

notandyvee committed:
[Bug] Fixed a coroutine bug and added error dialog for clarity.

notandyvee commented on Simplenote/src/main/java/com/automattic/simplenote/networking/SimpleHttp.kt:
```diff
@@ -0,0 +1,57 @@
+import org.json.JSONObject
+import java.util.concurrent.TimeUnit
+
+object SimpleHttp {
```
Just a heads up. While trying to add the unit tests I began to run into problems. Firstly, mocking static methods is annoying due to how behind we are with library updates. So I will have to make it a regular class. I will move `OkHttp` into a singleton instance using hilt so it can be injected into an `SimpleHttp` instance. Then we'll also inject `SimpleHttp`. This will make unit testing easier. I'll do it monday morning.

There is the other issue of passing `Simplenote`, which I cannot mock with the current version of mockito. I will try and also inject that in somehow through an interface. I have some thoughts. Either way, with the exception of those gotchas feel free to review the rest of it, which I will likely not change.

notandyvee committed:
Refactored in order to add unit tests.

Namely added a new networking module to be injected so we can
test a bit easier. Had to update mockito as well because it
couldn't mock basic classes.

notandyvee committed:
Add use of query const in DeeplinkActivity

notandyvee commented:
Ok @danilo04 , ready for review. Sorry it took so long. I had trouble with unit tests. I attempted to migrate to flows as a result, but I don't feel comfortable using outdated libraries. I tried updating a few libs and it caused a cascading effect of updating way too many things. We can save it for another time. Just wanted to be clear why I didn't use flows.

Additionally, in order to test this end-to-end, you must build it in release mode with the right creds. DM me if you need help. Won't work in debug at the moment due to the creds and debug package name.

Also, do I need to do anything special with strings? It's giving me a warning. But I'm pretty sure I only edited the base `strings.xml`.

roundhill commented on Simplenote/src/main/java/com/automattic/simplenote/viewmodels/RequestMagicLinkViewModel.kt:
```diff
@@ -0,0 +1,49 @@
+    val magicLinkRequestUiState: LiveData<MagicLinkRequestUiState> get() = _magicLinkRequestUiState
+
+    fun requestLogin(username: String) = viewModelScope.launch(ioDispatcher) {
+        _magicLinkRequestUiState.postValue(MagicLinkRequestUiState.Loading(message = "Requesting Magic Link"))
```
Can you make this localizable, and maybe say `Requesting log in link` instead?

notandyvee replied:
Yes! I guess I will do the same for the others. There are a few of those 😅 . Thanks for pointing it out.

notandyvee replied:
Done

notandyvee committed:
Removed use of raw string and use string res, plus add invalid link message

danilo04 commented on Simplenote/src/main/java/com/automattic/simplenote/DeepLinkActivity.kt:
```diff
@@ -0,0 +1,73 @@
+            intent.putExtra(SimplenoteAuthenticationActivity.KEY_MAGIC_LINK_AUTH_CODE, authCode)
+            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK)
+            startActivity(intent)
+        }
```
I think we can add a Toast message here for the else case, saying that the link was invalid. Just use the app context to launch the Toast.

danilo04 replied:
And take the user to the notes activity.

notandyvee replied:
Assuming no other big issues, I can make an update to this on a subsequent PR if that's ok.

danilo04 replied:
Yeah, this is a pretty edge case, but we cannot be sure if the email client will mess up the link or something like that. So if the link is wrong, then we need to send the user to the login screen if it is not logged in.

notandyvee replied:
> So if the link is wrong, then we need to send the user to the login screen if it is not logged in.

Oh. I see what you mean 🤔 . Fair enough. Created the issue.

danilo04 commented on Simplenote/src/main/java/com/automattic/simplenote/authentication/SignInFragment.kt:
```diff
@@ -0,0 +1,65 @@
+@AndroidEntryPoint
+class SignInFragment: MagicLinkableFragment() {
+
+    var viewModel: RequestMagicLinkViewModel? = null
```
You can use the property delegate to build the view model:

```suggestion
    val viewModel: RequestMagicLinkViewModel by viewModels()
```

danilo04 commented on Simplenote/src/main/java/com/automattic/simplenote/authentication/SignInFragment.kt:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+        viewModel = ViewModelProvider(this).get(RequestMagicLinkViewModel::class.java)
```
This can be deleted if you use the property delegate `by viewModels()` (comment above).

danilo04 commented on Simplenote/src/main/java/com/automattic/simplenote/networking/SimpleHttp.kt:
```diff
@@ -0,0 +1,50 @@
+
+private const val HTTP_SCHEME = "https"
+private const val BASE_URL = "app.simplenote.com"
+open class SimpleHttp @Inject constructor(val client: OkHttpClient) {
```
Consider refactoring `AccountNetworkUtils` by maybe using this class instead of the OkHtttp service it creates. It can be handled in a future PR.

notandyvee replied:
Oh hey. I guess you had a similar idea 😅 . Ya. Definitely future PR.

danilo04 commented on Simplenote/src/main/java/com/automattic/simplenote/viewmodels/CompleteMagicLinkViewModel.kt:
```diff
@@ -0,0 +1,62 @@
+                    val syncToken = json.getString("sync_token")
+                    val username = json.getString("username")
+                    simplenote.loginWithToken(username, syncToken)
+                } else if (response.code() == 400) {
+                    _magicLinkUiState.postValue(MagicLinkUiState.Error(messageRes = R.string.magic_link_complete_login_error_message))
```
(❤️ ) Cool

danilo04 commented on Simplenote/src/main/java/com/automattic/simplenote/viewmodels/CompleteMagicLinkViewModel.kt:
```diff
@@ -0,0 +1,62 @@
+    fun completeLogin(authKey: String, authCode: String) = viewModelScope.launch(ioDispatcher) {
+        _magicLinkUiState.postValue(MagicLinkUiState.Loading(messageRes = R.string.magic_link_complete_login_loading_message))
+        try {
+            simpleHttp.firePostRequest(
+                "account/complete-login",
+                mapOf(Pair("auth_key", authKey), Pair("auth_code", authCode))
+            ).use { response ->
```
Consider using a repository for this kind of call. Maybe you can create a `MagicLinkRepository` to handle the calls and then inject it into the view models that need it.

danilo04 commented on Simplenote/src/main/java/com/automattic/simplenote/viewmodels/RequestMagicLinkViewModel.kt:
```diff
@@ -0,0 +1,53 @@
+        )
+        try {
+            // TODO: Uncomment the request_source once finalized.
+            simpleHttp.firePostRequest(
```
Same comment as above, consider moving this call into a repository.

danilo04 commented on Simplenote/src/test/java/com/automattic/simplenote/viewmodels/CompleteMagicLinkViewModelTest.kt:
```diff
@@ -0,0 +1,78 @@
+const val AUTH_CODE_TEST = "auth_code_test"
+
+@ExperimentalCoroutinesApi
+class CompleteMagicLinkViewModelTest {
```
(❤️ ) Nice!

danilo04 reviewed (changes requested):
Hi @notandyvee, great job on the changes. The magic link is working as expected. However, it looks like it messed up the login with password. Check the following video:


https://github.com/Automattic/simplenote-android/assets/195721/4cfd8f52-60d5-485b-a362-dcad84d5906e

Also, if I try to open a magic link and the user is already logged in, the user is logged out. We need to handle this edge case. 

I left a couple of code comments for your consideration as well.

notandyvee committed:
[Bug] Fix bug where magic link logs you out if already logged in.

The fix for now is to avoid making the magic link calls and instead
go to the notes activity. Basically using the existing behavior.

notandyvee committed:
[Bug] Fixed bug that made login with password seem broken.

notandyvee committed:
Use repository instead of OkHttp libs directly.

notandyvee commented:
@danilo04 Ready for re-review. I will not add unit tests for the 2 classes the build was complaining about since it would essentially be testing library code. So added the unit test exemption. Appreciate the thorough review.

notandyvee committed:
Updated release notes.

danilo04 reviewed (approved):
Hi @notandyvee, thank you for the changes. The PR is looking good and I tested the scenarios that were failing and now they are working. Good job on this :shipit: 

We can handle the edge case discussed in the code comments in a different PR.
